src/Params/PADnoteParameters.cpp

```diff
@@ -935,54 +967,129 @@ int PADnoteParameters::sampleGenerator(PADnoteParameters::callback cb,
         FFTwrapper *fft      = new FFTwrapper(samplesize);
         fft_t      *fftfreqs = new fft_t[samplesize / 2];
         float      *spectrum = new float[spectrumsize];
+        const OscilGen* oscilgen_ptr = this_c->oscilgen;
 
         for(int nsample = 0; nsample < samplemax; ++nsample)
         if(nsample % nthreads == threadno)
         {
-            if(do_abort())
-                break;
+            assert(oscilgen_ptr);
+#if 0
+            // UNUSED
+            // compute min and max for the wavetable
+            // no float comparisons here to avoid comparison errors
+            int max_extent; // in range [0,63]
+            
+            if(this_c->Pmode == pad_mode::wavetable)
+            {
+                max_extent = std::min((int)oscilgen_ptr->Pbasefuncpar, 127-oscilgen_ptr->Pbasefuncpar);
+                if(max_extent < 0) { max_extent = 0; }
+            }
+            else
+            {
+                max_extent = 0;
+            }
+
+            std::size_t minIdx = (oscilgen_ptr->Pbasefuncpar - max_extent) * Sample::num_smps_per_key;
+            std::size_t maxIdx = (oscilgen_ptr->Pbasefuncpar + max_extent) * Sample::num_smps_per_key + 1;
+            assert(minIdx < maxIdx);
+            assert(minIdx > 0);
+            assert(maxIdx <= 512);
+            assert(minIdx <= oscilgen_ptr->Pbasefuncpar*Sample::num_smps_per_key + 1.0f);
+            assert(maxIdx >= oscilgen_ptr->Pbasefuncpar*Sample::num_smps_per_key - 1.0f);
+#else
+            std::size_t minIdx, maxIdx;
```

Comment:
So, the idea with Pad Synth is to create a 2D matrix where dimension one is time, dimension two is expected sampling frequency. There's some randomization so typically you can't trivially walk between the different sampling freq rows.

I would have expected wavetable synthesis to turn this into a 3D tensor where you have timeXnote-freqXnote-parameter. It looks like instead there's just timeXnote-parameter here.

Reply:
Before this commit, we had `Sample sample[PAD_MAX_SAMPLES]` (freqs), and each `Sample` had one `float* smp` (array for time). Now, each `Sample` contains `std::vector<float*> smp` instead of `float* smp` - each vector element for one wavetable-parameter. So it's indeed 3D now.